File: ECHA.API/Controllers/ModeracaoController.cs
using EconomiaComHistoria.Core.DTOs;
using EconomiaComHistoria.Core.Enums;
using EconomiaComHistoria.Core.Interfaces;
using EconomiaComHistoria.Infrastructure.Data;
using EconomiaComHistoria.Infrastructure.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace EconomiaComHistoria.API.Controllers;

[ApiController]
[Authorize(Roles = "Admin,Editor,Moderador,SuperAdmin")]
[Route("api/moderacao")]
public class ModeracaoController : ControllerBase
{
    private readonly AppDbContext _dbContext;
    private readonly INotificacaoService _notificacaoService;
    private readonly IGamificacaoService _gamificacaoService;

    public ModeracaoController(AppDbContext dbContext, INotificacaoService notificacaoService, IGamificacaoService gamificacaoService)
    {
        _dbContext = dbContext;
        _notificacaoService = notificacaoService;
        _gamificacaoService = gamificacaoService;
    }

    [HttpGet("pendentes")]
    public async Task<ActionResult<ModeracaoPendentesResponse>> GetPendentes(CancellationToken cancellationToken)
    {
        var topicos = await _dbContext.TopicosForum
            .Include(x => x.Autor)
            .Include(x => x.Categoria)
            .Where(x => x.Estado == EstadoTopicoForum.Pendente)
            .OrderBy(x => x.CriadoEm)
            .Select(x => new ModeracaoPendenteDto(
                x.Id,
                "Topico",
                x.Titulo,
                x.AutorId,
                x.Autor == null ? null : x.Autor.Nome,
                x.CriadoEm,
                x.CategoriaId,
                x.Categoria == null ? null : x.Categoria.Nome,
                null,
                x.Denuncias.Count
            ))
            .ToListAsync(cancellationToken);

        var respostas = await _dbContext.RespostasForum
            .Include(x => x.Autor)
            .Where(x => x.EstadoResposta == EstadoComentario.Pendente)
            .OrderBy(x => x.DataCriacao)
            .Select(x => new ModeracaoPendenteDto(
                x.Id,
                "Resposta",
                x.Conteudo,
                x.AutorId,
                x.Autor == null ? null : x.Autor.Nome,
                x.DataCriacao,
                null,
                null,
                x.TopicoId,
                x.Denuncias.Count
            ))
            .ToListAsync(cancellationToken);

        return Ok(new ModeracaoPendentesResponse(topicos, respostas));
    }

    [HttpGet("denuncias")]
    public async Task<ActionResult<IEnumerable<DenunciaSummaryDto>>> GetDenuncias(CancellationToken cancellationToken)
    {
        var topicos = await _dbContext.TopicosForum
            .Include(x => x.Autor)
            .Where(x => x.Denuncias.Any())
            .OrderByDescending(x => x.Denuncias.Count)
            .Select(x => new DenunciaSummaryDto(
                x.Id,
                "Topico",
                x.Titulo,
                x.AutorId,
                x.Autor == null ? null : x.Autor.Nome,
                x.Denuncias.Count,
                x.Denuncias.Max(d => d.DataDenuncia)
            ))
            .ToListAsync(cancellationToken);

        return Ok(topicos);
    }

    [HttpGet("utilizadores")]
    public async Task<ActionResult<IEnumerable<UtilizadorModeracaoDto>>> GetUtilizadores(CancellationToken cancellationToken)
    {
        var utilizadores = await _dbContext.Utilizadores
            .Select(x => new UtilizadorModeracaoDto(
                x.Id,
                x.Nome,
                x.Email,
                x.Tipo.ToString(),
                x.Suspenso,
                x.SuspensoAte,
                x.SuspensaoPermanente
            ))
            .ToListAsync(cancellationToken);

        return Ok(utilizadores);
    }

    [HttpPut("topicos/{id:int}/aprovar")]
    public async Task<IActionResult> AprovarTopico(int id, CancellationToken cancellationToken)
    {
        var topico = await _dbContext.TopicosForum.FindAsync(new object[] { id }, cancellationToken);
        if (topico is null)
            return NotFound(new { message = "Topico nao encontrado" });

        topico.Estado = EstadoTopicoForum.Ativo;
        await _dbContext.SaveChangesAsync(cancellationToken);

        await _notificacaoService.EnviarPushAsync(topico.AutorId, "Topico aprovado", $"O topico \"{topico.Titulo}\" foi aprovado.", cancellationToken);
        return NoContent();
    }

    [HttpPut("respostas/{id:int}/aprovar")]
    public async Task<IActionResult> AprovarResposta(int id, CancellationToken cancellationToken)
    {
        var resposta = await _dbContext.RespostasForum.FindAsync(new object[] { id }, cancellationToken);
        if (resposta is null)
            return NotFound(new { message = "Resposta nao encontrada" });

        resposta.EstadoResposta = EstadoComentario.Aprovada;
        await _dbContext.SaveChangesAsync(cancellationToken);

        await _notificacaoService.EnviarPushAsync(resposta.AutorId, "Resposta aprovada", "A sua resposta foi aprovada.", cancellationToken);
        return NoContent();
    }

    [HttpPut("topicos/{id:int}/rejeitar")]
    public async Task<IActionResult> RejeitarTopico(int id, [FromBody] RejeitarTopicoDto request, CancellationToken cancellationToken)
    {
        var topico = await _dbContext.TopicosForum.FindAsync(new object[] { id }, cancellationToken);
        if (topico is null)
            return NotFound(new { message = "Topico nao encontrado" });

        topico.Estado = EstadoTopicoForum.Rejeitado;
        await _dbContext.SaveChangesAsync(cancellationToken);

        await _notificacaoService.EnviarPushAsync(topico.AutorId, "Topico rejeitado", request.MotivoRejeicao, cancellationToken);
        return NoContent();
    }

    [HttpPut("respostas/{id:int}/rejeitar")]
    public async Task<IActionResult> RejeitarResposta(int id, [FromBody] RejeitarTopicoDto request, CancellationToken cancellationToken)
    {
        var resposta = await _dbContext.RespostasForum.FindAsync(new object[] { id }, cancellationToken);
        if (resposta is null)
            return NotFound(new { message = "Resposta nao encontrada" });

        resposta.EstadoResposta = EstadoComentario.Removido;
        await _dbContext.SaveChangesAsync(cancellationToken);

        await _notificacaoService.EnviarPushAsync(resposta.AutorId, "Resposta rejeitada", request.MotivoRejeicao, cancellationToken);
        return NoContent();
    }

    [HttpGet("denuncias/detalhado")]
    public async Task<ActionResult> GetDenunciasDetalhado(CancellationToken cancellationToken)
    {
        var denuncias = await _dbContext.Denuncias
            .Include(x => x.TopicoForum)
                .ThenInclude(t => t!.Autor)
            .Include(x => x.RespostaForum)
                .ThenInclude(c => c!.Autor)
            .OrderByDescending(x => x.DataDenuncia)
            .Select(x => new
            {
                x.Id,
                x.TipoAlvo,
                x.IdAlvo,
                x.Motivo,
                x.Descricao,
                x.Estado,
                x.DataDenuncia,
                Topico = x.TopicoForum != null ? new
                {
                    x.TopicoForum.Id,
                    x.TopicoForum.Titulo,
                    AutorNome = x.TopicoForum.Autor != null ? x.TopicoForum.Autor.Nome : null
                } : null,
                Resposta = x.RespostaForum != null ? new
                {
                    x.RespostaForum.Id,
                    x.RespostaForum.Conteudo,
                    AutorNome = x.RespostaForum.Autor != null ? x.RespostaForum.Autor.Nome : null
                } : null
            })
            .ToListAsync(cancellationToken);

        return Ok(denuncias);
    }

    [HttpPut("utilizadores/{id:int}/suspender")]
    public async Task<IActionResult> SuspenderUtilizador(int id, [FromBody] SuspenderUtilizadorDto request, CancellationToken cancellationToken)
    {
        var utilizador = await _dbContext.Utilizadores.FindAsync(new object[] { id }, cancellationToken);
        if (utilizador is null)
            return NotFound(new { message = "Utilizador nao encontrado" });

        utilizador.SuspensoAte = request.DiasSuspensao.HasValue
            ? DateTime.UtcNow.AddDays(request.DiasSuspensao.Value)
            : DateTime.UtcNow.AddYears(100);

        await _dbContext.SaveChangesAsync(cancellationToken);
        return NoContent();
    }

    [HttpPut("utilizadores/{id:int}/reativar")]
    public async Task<IActionResult> ReativarUtilizador(int id, CancellationToken cancellationToken)
    {
        var utilizador = await _dbContext.Utilizadores.FindAsync(new object[] { id }, cancellationToken);
        if (utilizador is null)
            return NotFound(new { message = "Utilizador nao encontrado" });

        utilizador.SuspensoAte = null;

        await _dbContext.SaveChangesAsync(cancellationToken);
        return NoContent();
    }

    [HttpGet("badges")]
    public async Task<ActionResult<List<BadgeAdminDto>>> GetBadges(CancellationToken cancellationToken)
    {
        var badges = await _dbContext.Badges
            .Include(b => b.Conquistado)
            .Select(b => new BadgeAdminDto(
                b.Id,
                b.Nome,
                b.Descricao,
                b.IconeUrl,
                b.CriterioTipo,
                b.CriterioValor,
                b.Conquistado.Count
            ))
            .ToListAsync(cancellationToken);
        return Ok(badges);
    }

    [HttpGet("metricas-engajamento")]
    public async Task<ActionResult<object>> GetMetricasEngajamento(CancellationToken cancellationToken)
    {
        var metricas = await _gamificacaoService.GetMetricasEngajamentoAsync(cancellationToken);
        return Ok(metricas);
    }
}

File: ECHA.Web/Controllers/ModeracaoController.fs
namespace ECHA.Web.Controllers

open System
open System.Collections.Generic
open System.Threading.Tasks
open Microsoft.AspNetCore.Mvc
open Microsoft.AspNetCore.Authorization
open Microsoft.AspNetCore.Http
open EconomiaComHistoria.Core.DTOs
open ECHA.Web.Services

[<Authorize(Roles = "Admin,Editor,Moderador,SuperAdmin")>]
type ModeracaoController (apiClient: ECHA.Web.Services.ApiClient) =
    inherit Controller()

    member private this.GetToken() =
        let claim = this.User.FindFirst("AccessToken")
        if isNull claim then null else claim.Value

    [<HttpGet>]
    member this.Fila () =
        task {
            let token = this.GetToken()
            match token with
            | null -> return this.Unauthorized() :> IActionResult
            | t ->
                let! pendentes = apiClient.GetPendentesAsync(t)
                match pendentes with
                | Some p -> return this.View(p) :> IActionResult
                | None -> 
                    let listaVazia1 = System.Collections.Generic.List<ModeracaoPendenteDto>()
                    let listaVazia2 = System.Collections.Generic.List<ModeracaoPendenteDto>()
                    return this.View(ModeracaoPendentesResponse(listaVazia1, listaVazia2)) :> IActionResult
        }

    [<HttpGet>]
    member this.Denuncias () =
        task {
            let token = this.GetToken()
            match token with
            | null -> return this.Unauthorized() :> IActionResult
            | t ->
                let! denuncias = apiClient.GetDenunciasAsync(t)
                return this.View(denuncias) :> IActionResult
        }

    [<HttpGet>]
    member this.Utilizadores () =
        task {
            let token = this.GetToken()
            match token with
            | null -> return this.Unauthorized() :> IActionResult
            | t ->
                let! utilizadores = apiClient.ListUtilizadoresAsync(t)
                return this.View(utilizadores) :> IActionResult
        }

    [<HttpPost>]
    [<ValidateAntiForgeryToken>]
    member this.AprovarTopico (id: int) =
        task {
            let token = this.GetToken()
            match token with
            | null -> 
                this.TempData["ErrorMessage"] <- "Sessão expirada. Faça login novamente."
                return this.RedirectToAction("Login", "Auth") :> IActionResult
            | t ->
                try
                    let! sucesso = apiClient.AprovarTopicoAsync(id, t)
                    if sucesso then
                        this.TempData["SuccessMessage"] <- "Tópico aprovado com sucesso."
                    else
                        this.TempData["ErrorMessage"] <- "Falha ao aprovar tópico."
                with
                | :? ApiClientException as ex ->
                    this.TempData["ErrorMessage"] <- $"Erro: {ex.Message}"
                return this.RedirectToAction("Denuncias") :> IActionResult
        }

    [<HttpPost>]
    [<ValidateAntiForgeryToken>]
    member this.RejeitarTopico (id: int, motivo: string) =
        task {
            let token = this.GetToken()
            match token with
            | null -> 
                this.TempData["ErrorMessage"] <- "Sessão expirada. Faça login novamente."
                return this.RedirectToAction("Login", "Auth") :> IActionResult
            | t ->
                try
                    let request = RejeitarTopicoDto(motivo)
                    let! sucesso = apiClient.RejeitarTopicoAsync(id, request, t)
                    if sucesso then
                        this.TempData["SuccessMessage"] <- "Tópico rejeitado e penalização aplicada."
                    else
                        this.TempData["ErrorMessage"] <- "Falha ao rejeitar tópico."
                with
                | :? ApiClientException as ex ->
                    this.TempData["ErrorMessage"] <- $"Erro: {ex.Message}"
                return this.RedirectToAction("Denuncias") :> IActionResult
        }

    [<HttpPost>]
    [<ValidateAntiForgeryToken>]
    member this.AprovarResposta (id: int) =
        task {
            let token = this.GetToken()
            match token with
            | null -> 
                this.TempData["ErrorMessage"] <- "Sessão expirada. Faça login novamente."
                return this.RedirectToAction("Login", "Auth") :> IActionResult
            | t ->
                try
                    let! sucesso = apiClient.AprovarRespostaAsync(id, t)
                    if sucesso then
                        this.TempData["SuccessMessage"] <- "Resposta aprovada com sucesso."
                    else
                        this.TempData["ErrorMessage"] <- "Falha ao aprovar resposta."
                with
                | :? ApiClientException as ex ->
                    this.TempData["ErrorMessage"] <- $"Erro: {ex.Message}"
                return this.RedirectToAction("Denuncias") :> IActionResult
        }

    [<HttpPost>]
    [<ValidateAntiForgeryToken>]
    member this.RejeitarResposta (id: int, motivo: string) =
        task {
            let token = this.GetToken()
            match token with
            | null -> 
                this.TempData["ErrorMessage"] <- "Sessão expirada. Faça login novamente."
                return this.RedirectToAction("Login", "Auth") :> IActionResult
            | t ->
                try
                    let request = RejeitarTopicoDto(motivo)
                    let! sucesso = apiClient.RejeitarRespostaAsync(id, request, t)
                    if sucesso then
                        this.TempData["SuccessMessage"] <- "Resposta rejeitada e penalização aplicada."
                    else
                        this.TempData["ErrorMessage"] <- "Falha ao rejeitar resposta."
                with
                | :? ApiClientException as ex ->
                    this.TempData["ErrorMessage"] <- $"Erro: {ex.Message}"
                return this.RedirectToAction("Denuncias") :> IActionResult
        }

    [<HttpPost>]
    [<ValidateAntiForgeryToken>]
    member this.SuspenderUtilizador (id: int, dias: Nullable<int>) =
        task {
            let token = this.GetToken()
            match token with
            | null -> 
                this.TempData["ErrorMessage"] <- "Sessão expirada."
                return this.RedirectToAction("Login", "Auth") :> IActionResult
            | t ->
                try
                    let request = SuspenderUtilizadorDto(dias)
                    let! sucesso = apiClient.SuspenderUtilizadorAsync(id, request, t)
                    if sucesso then
                        this.TempData["SuccessMessage"] <- "Utilizador suspenso com sucesso."
                    else
                        this.TempData["ErrorMessage"] <- "Falha ao suspender utilizador."
                with
                | :? ApiClientException as ex ->
                    this.TempData["ErrorMessage"] <- $"Erro: {ex.Message}"
                return this.RedirectToAction("Utilizadores") :> IActionResult
        }

    [<HttpPost>]
    [<ValidateAntiForgeryToken>]
    member this.ReativarUtilizador (id: int) =
        task {
            let token = this.GetToken()
            match token with
            | null -> 
                this.TempData["ErrorMessage"] <- "Sessão expirada."
                return this.RedirectToAction("Login", "Auth") :> IActionResult
            | t ->
                try
                    let! sucesso = apiClient.ReativarUtilizadorAsync(id, t)
                    if sucesso then
                        this.TempData["SuccessMessage"] <- "Utilizador reativado com sucesso."
                    else
                        this.TempData["ErrorMessage"] <- "Falha ao reativar utilizador."
                with
                | :? ApiClientException as ex ->
                    this.TempData["ErrorMessage"] <- $"Erro: {ex.Message}"
                return this.RedirectToAction("Utilizadores") :> IActionResult
        }